### PR TITLE
Masses -- implementation as described in #807

### PR DESCRIPTION
**Pull request by hjr**
Brief summary of the changes
----------------------------
Started to work on code for #807. The pull request is more ment to run robots on it and have a decent review, since this is touching the core xcs functionalities.
- Split dry mass into empty and crew mass.
- empty mass is stored in the plane profile. It is a attribute easily readable from cockpit or glider manual (wight balance report).
- crew mass is volatile and part of the flight configuration, whereas a default can now be entered to the pilot identity found for the logger. (open for suggestions here)

Related issues and discussions
------------------------------
Addresses #807

### CODE REVIEW

**Comment by hjr** [comment]
Of course, unit test got updated.

**Comment by MaxKellermann** [comment]
No merge commits in PRs, please

**Comment by hjr** [comment]
Next step on this branch is to find a solution for a default for the empty weight of the glider.
 - recalc all polar data to a reference weight that can be used as a default for the empty weight.
 - add an empty weight default to the polar store
 - mazbe drop the 3 point storage format in favor for an a,b,c polynom factor format. just display 3 points
let's see what is working best.

**Comment by hjr** [comment]
Yip, can hash the merge commit out ...

File: src/Dialogs/Plane/PlanePolarDialog.cpp [hunk 1]
```diff
@@ -39,14 +39,16 @@ Copyright_License {
 #include "Language/Language.hpp"
 #include "UIGlobals.hpp"
 
+#include "cstring"
```
Comment (MaxKellermann) [inline src/Dialogs/Plane/PlanePolarDialog.cpp:42]: Why this include?
External headers should be included with `<cstring>` not `"cstring"`.

File: src/Dialogs/Plane/PlanePolarDialog.cpp [hunk 2]
```diff
@@ -132,15 +134,15 @@ PlanePolarWidget::Prepare(ContainerWindow &parent,
   DataFieldListener *listener = this;
   Add(std::make_unique<PolarShapeEditWidget>(plane.polar_shape, listener));
 
-  AddFloat(_("Reference Mass"), _("Reference mass of the polar"),
+  AddFloat(_("Reference Mass"), _("Reference mass of the polar."),
```
Comment (MaxKellermann) [inline src/Dialogs/Plane/PlanePolarDialog.cpp:137]: Unrelated change, doesn't belong in this commit

File: src/Dialogs/Plane/PlanePolarDialog.cpp [hunk 3]
```diff
@@ -187,8 +189,9 @@ PlanePolarWidget::ListClicked()
 
   const PolarStore::Item &item = PolarStore::GetItem(list[result].int_value);
 
-  plane.reference_mass = item.reference_mass;
-  plane.dry_mass = item.reference_mass;
+  plane.polar_shape.reference_mass = item.reference_mass;
+  auto crew_mass_estimate = StaticString<32>(item.name).Contains(_T("PAS)")) ? 180. : 90.;
```
Comment (MaxKellermann) [inline src/Dialogs/Plane/PlanePolarDialog.cpp:193]: I don't like matching a human-readable name for such "magic" sub strings. This is error prone. Never parse human-readable strings, please.
Also calling `StaticString` creates a truncated copy on the stack, and I don't know why you did that.

File: src/Dialogs/Plane/PlanePolarDialog.cpp [hunk 4]
```diff
@@ -187,8 +189,9 @@ PlanePolarWidget::ListClicked()
...
+  plane.empty_mass = item.reference_mass - crew_mass_estimate; // todo needs rework on polar store data
```
Comment (MaxKellermann) [inline src/Dialogs/Plane/PlanePolarDialog.cpp:194]: Can we solve this todo before merging this PR?

File: src/Dialogs/Settings/Panels/LoggerConfigPanel.cpp [hunk 5]
```diff
@@ -28,15 +28,19 @@ Copyright_License {
 #include "Widget/RowFormWidget.hpp"
 #include "UIGlobals.hpp"
 #include "Form/DataField/Enum.hpp"
+#include "Form/DataField/Float.hpp"
 #include "Logger/NMEALogger.hpp"
 #include "UtilsSettings.hpp"
 #include "Components.hpp"
+#include "Units/Units.hpp"
```
Comment (MaxKellermann) [inline src/Dialogs/Settings/Panels/LoggerConfigPanel.cpp:35]: Unnecessary include

File: src/Dialogs/Settings/Panels/LoggerConfigPanel.cpp [hunk 6]
```diff
@@ -28,15 +28,19 @@ Copyright_License {
 #include "Widget/RowFormWidget.hpp"
 #include "UIGlobals.hpp"
 #include "Form/DataField/Enum.hpp"
+#include "Form/DataField/Float.hpp"
```
Comment (MaxKellermann) [inline src/Dialogs/Settings/Panels/LoggerConfigPanel.cpp:31]: Unnecessary include

File: src/Engine/GlideSolvers/GlidePolar.hpp [hunk 7]
```diff
@@ -66,7 +66,7 @@ class GlidePolar
 
   /** Clean ratio (1=clean, 0=100% bugs) */
   double bugs;
-  /** Ballast ratio (litres) */
+  /** Ballast (litres alias kg) */
```
Comment (MaxKellermann) [inline src/Engine/GlideSolvers/GlidePolar.hpp:69]: This change is good, but is unrelated to this commit; actually, we should better never refer to "ballast litres" at all, because the ballast volume is not relevant for anything.

File: src/Plane/PlaneFileGlue.cpp [hunk 8]
```diff
@@ -128,8 +132,13 @@ PlaneGlue::Read(Plane &plane, KeyValueFileReader &reader)
     plane.weglide_glider_type = 0;
   if (!has_polar_name)
     plane.polar_name.clear();
-  if (!has_dry_mass)
-    plane.dry_mass = plane.reference_mass;
+  if (!has_empty_mass && has_dry_mass) {
+    auto crew_mass_estimate = plane.polar_name.Contains(_T("PAS)")) ? 180. : 90.;
```
Comment (MaxKellermann) [inline src/Plane/PlaneFileGlue.cpp:136]: Another match in a human-readable string. Please don't.

File: src/Plane/PlaneFileGlue.cpp [hunk 9]
```diff
@@ -172,10 +181,10 @@ PlaneGlue::Write(const Plane &plane, KeyValueFileWriter &writer)
   FormatPolarShape(plane.polar_shape, tmp.buffer(), tmp.capacity());
   writer.Write("PolarInformation", tmp);
 
-  tmp.Format("%f", (double)plane.reference_mass);
+  tmp.Format("%f", (double)plane.polar_shape.reference_mass);
   writer.Write("PolarReferenceMass", tmp);
-  tmp.Format("%f", (double)plane.dry_mass);
-  writer.Write("PolarDryMass", tmp);
```
Comment (MaxKellermann) [inline src/Plane/PlaneFileGlue.cpp:178]: I think it would be better to write both for some time, to keep some compatibility with old versions, just in case somebody downgrades XCSoar for some reason. It should be accompanied with a good code comment.

File: test/data/D-4449.xcp [hunk 10]
```diff
@@ -5,7 +5,7 @@ Handicap="100"
 PolarName="Hornet"
 PolarInformation="80.000,-0.606,120.000,-0.990,160.000,-1.918"
 PolarReferenceMass="318"
-PolarDryMass="302"
+PlaneEmptyMass="302"
```
Comment (MaxKellermann) [inline test/data/D-4449.xcp:8]: Can you please add another unit test which ensures that reading plane files from old XCSoar versions keeps working? I don't want to break this, this is important.

**Review by MaxKellermann — CHANGES_REQUESTED** [review]

File: src/Dialogs/Plane/PlanePolarDialog.cpp (hunk 3 above)
Comment (hjr) [reply src/Dialogs/Plane/PlanePolarDialog.cpp:193]: See below.

File: src/Dialogs/Plane/PlanePolarDialog.cpp (hunk 4 above)
Comment (hjr) [reply src/Dialogs/Plane/PlanePolarDialog.cpp:194]: Yes, no way around that. The commit is according to me UML sketch, but not complete. Just going forward not having a buy in would be a mess.

File: src/Dialogs/Settings/Panels/LoggerConfigPanel.cpp (hunk 5 above)
Comment (hjr) [reply src/Dialogs/Settings/Panels/LoggerConfigPanel.cpp:35]: Ok

File: src/Dialogs/Settings/Panels/LoggerConfigPanel.cpp (hunk 6 above)
Comment (hjr) [reply src/Dialogs/Settings/Panels/LoggerConfigPanel.cpp:31]: ok

File: src/Engine/GlideSolvers/GlidePolar.hpp (hunk 7 above)
Comment (hjr) [reply src/Engine/GlideSolvers/GlidePolar.hpp:69]: It is related. The commit/branch tries to clean up the masses mess and unclear code comments are a target.

File: src/Plane/PlaneFileGlue.cpp (hunk 8 above)
Comment (hjr) [reply src/Plane/PlaneFileGlue.cpp:136]: Yip, as above.

File: src/Plane/PlaneFileGlue.cpp (hunk 9 above)
Comment (hjr) [reply src/Plane/PlaneFileGlue.cpp:178]: Good hint.

File: test/data/D-4449.xcp (hunk 10 above)
Comment (hjr) [reply test/data/D-4449.xcp:8]: Good hint.

**Comment by hjr** [comment]
Working through the PIL/PAS hurdles to set a proper empty mass for the all the gliders in the polar store I quite often run into such things. It make you think xcs might just be a useless toy. (PIL/PAS polars mapped to the same mass should just perfectly overlay).
DG1000_PILPAS

File: src/Dialogs/Plane/PlanePolarDialog.cpp (hunk 1 above)
Comment (hjr) [reply src/Dialogs/Plane/PlanePolarDialog.cpp:42]: removed

File: src/Dialogs/Plane/PlanePolarDialog.cpp (hunk 2 above)
Comment (hjr) [reply src/Dialogs/Plane/PlanePolarDialog.cpp:137]: see below

Comment (hjr) [reply src/Dialogs/Plane/PlanePolarDialog.cpp:137]: see below

File: src/Plane/PlaneFileGlue.cpp (hunk 9 above)
Comment (hjr) [reply src/Plane/PlaneFileGlue.cpp:178]: Old dry_mass kept in file and structure, keeping it up to date just for a potential downgrade of the xcs.

File: src/Dialogs/Plane/PlanePolarDialog.cpp (hunk 4 above)
Comment (hjr) [reply src/Dialogs/Plane/PlanePolarDialog.cpp]: Removed this logic with the intention to commit the reorganized polar store in a second commit incl. the proper empty mass default..

**Comment by hjr** [comment]
Thx for the review, Max!

**Comment by hjr** [comment]
There are unfortunately some more of this kind. I just drop the obviously wrong one.
is-28B2

**Comment by hjr** [comment]
Guess I worked around the official process of addressing review findings by force pushing the fixes to them .. hope you are forgiving on this.

File: src/Dialogs/Plane/PlanePolarDialog.cpp (hunk 2 above)
Comment (MaxKellermann) [reply src/Dialogs/Plane/PlanePolarDialog.cpp:133]: I don't understand, see below what?

File: src/Dialogs/Settings/dlgBasicSettings.cpp [hunk 11]
```diff
@@ -271,9 +282,16 @@ FlightSetupPanel::Prepare(ContainerWindow &parent,
...
+           _("All masses loaded to the glider beyond the empty weight including pilot and copilot, but not water ballast."),
+           _T("%.0f %s"), _T("%.0f"),
+           0, 300, 5, false, UnitGroup::MASS,
+           polar_settings.glide_polar_task.GetCrewMass(),
+           this);
+  
   const double db = 5;
   AddFloat(_("Ballast"),
-           _("Ballast of the glider.  Increase this value if the pilot/cockpit load is greater than the reference pilot weight of the glide polar (typically 75kg).  Press ENTER on this field to toggle count-down of the ballast volume according to the dump rate specified in the configuration settings."),
+           _("Ballast of the glider. Press ENTER on this field to toggle count-down of the ballast volume according to the dump rate specified in the configuration settings."),
```
Comment (MaxKellermann) [inline src/Dialogs/Settings/dlgBasicSettings.cpp:294]: What does "ENTER" mean? The "ENTER" key? Most devices don't have an "ENTER" key; wording should be portable, not just the code.

File: src/Engine/GlideSolvers/GlidePolar.hpp (hunk 7 above)
Comment (MaxKellermann) [reply src/Engine/GlideSolvers/GlidePolar.hpp:69]: This commit is documented as "Split dry_mass into empty + crew mass, attached reference mass to polar shape", and this change is confusing - not because it is bad, but it does not "split dry_mass into empty+crew mass".

File: src/Plane/PlaneFileGlue.cpp (hunk 9 above)
Comment (MaxKellermann) [reply src/Plane/PlaneFileGlue.cpp:178]: There's no code comment explanining the deprecation. Writing one takes less than a minute, and will save the next guy an hour.

File: src/Polar/PolarStore.cpp [hunk 12]
```diff
@@ -58,256 +58,206 @@ PolarStore::Item::ToPolarInfo() const
 }
 
 /**
- *  Note: new items should be added to bottom, otherwise saved index is lost 
- *  updated 2013/02/04 OH DAEC list 2012
```
Comment (MaxKellermann) [inline src/Polar/PolarStore.cpp:62]: You must not reorder the items!
This code is really badly designed, but those polars are referenced by their index in this array.

**Review by MaxKellermann — CHANGES_REQUESTED** [review]

File: src/Dialogs/Settings/dlgBasicSettings.cpp (hunk 11 above)
Comment (hjr) [reply src/Dialogs/Settings/dlgBasicSettings.cpp:294]: Not introduced through my change. I am responsible for the remove of the first part of the help message, which is now valid for the crew load.

Comment (MaxKellermann) [reply src/Dialogs/Settings/dlgBasicSettings.cpp:294]: Oh, right! Sorry, this is an old bad message.

File: src/Polar/PolarStore.cpp (hunk 12 above)
Comment (hjr) [reply src/Polar/PolarStore.cpp:62]: Removed index access to polar store entirely.

**Comment by hjr** [comment]
https://github.com/XCSoar/XCSoar/pull/840#discussion_r788912598 Did not find the comment button to this. Me previous comment "below" was: It is related. The commit/branch tries to clean up the masses mess and unclear code comments are a target.

File: src/Dialogs/Settings/dlgBasicSettings.cpp (hunk 11 above)
Comment (hjr) [reply src/Dialogs/Settings/dlgBasicSettings.cpp]: Replaced the ENTER in the help text.

File: src/Plane/PlaneFileGlue.cpp (hunk 9 above)
Comment (hjr) [reply src/Plane/PlaneFileGlue.cpp:178]: Added one.

File: src/Dialogs/Plane/PlanePolarDialog.cpp (hunk 2 above)
Comment (hjr) [reply src/Dialogs/Plane/PlanePolarDialog.cpp:134]: My previous comment "below" was: It is related. The commit/branch tries to clean up the masses mess and unclear code comments are a target.

**Comment by MaxKellermann** [comment]
Yay, I'm glad you didn't give up. Stuff like this is thankless, tedious work, only the lack of which is ever noticed. Thanks.
